modules/core/shared/src/main/scala/io/circe/Cursor.scala

```diff
@@ -0,0 +1,1351 @@
+      override val obj: JsonObject,
+      override val keyValue: String,
+      override protected val unsafeLastCursor: Cursor.SuccessCursor,
+      override protected val unsafeLastOp: CursorOp,
+      override val parent: SuccessCursor,
+      override val changed: Boolean
+    ) extends ObjectCursor
+
+    private[Cursor] def apply(
+      obj: JsonObject,
+      keyValue: String,
+      unsafeLastCursor: Cursor.SuccessCursor,
+      unsafeLastOp: CursorOp,
+      parent: SuccessCursor,
+      changed: Boolean
+    ): ObjectCursor =
+      ObjectCursorImpl(obj, keyValue, unsafeLastCursor, unsafeLastOp, parent, changed)
+  }
+
+  /**
+   * A cursor in a failed state. The only way to exit this state is via the
+   * [[Cursor#lastCursor]] method.
+   */
+  private sealed abstract class FailedCursor extends Cursor {
+
+    /**
+     * FailedCursor values always have a failure reason.
+     */
+    def failureReasonValue: CursorFailureReason
+
+    override final def failureReason: Some[CursorFailureReason] = Some(failureReasonValue)
+    override final def focus: Option[Json] = None
+    override final def up: Cursor = this
+    override final def delete: Cursor = this
+    override final def index: Option[Int] = None
+    override final def key: Option[String] = None
+    override final def left: Cursor = this
+    override final def right: Cursor = this
+    override final def mapFocus(f: Json => Json): Cursor = this
+    override final def mapFocusA[F[_]](f: Json => F[Json])(implicit F: Applicative[F]): F[Cursor] = F.pure(this)
+    override final def downArray: Cursor = this
+    override final def downN(n: Int): Cursor = this
+    override final def field(k: String): Cursor = this
+    override final def downField(k: String): Cursor = this
+    override final def isRoot: Boolean = false
+
+    override final def toString: String =
+      s"FailedCursor(failureReason = $failureReason)"
+  }
+
+  private object FailedCursor {
+    private[this] final case class FailedCursorImpl(
+      override protected val unsafeLastCursor: Cursor.SuccessCursor,
+      override protected val unsafeLastOp: CursorOp,
+      override val failureReasonValue: CursorFailureReason
+    ) extends FailedCursor
+
+    private[Cursor] def apply(
+      unsafeLastCursor: Cursor.SuccessCursor,
+      unsafeLastOp: CursorOp,
+      failureReason: CursorFailureReason
+    ): FailedCursor =
+      FailedCursorImpl(unsafeLastCursor, unsafeLastOp, failureReason)
+  }
+
+  /**
+   * A representation of the cursor history. History instances can only be
+   * constructed from [[Cursor]] values. This is to ensure they always
+   * represent valid cursor historys.
+   */
+  sealed abstract class History extends Product with Serializable {
+
+    /** A representation of the [[History]] as a `List`. */
+    def asList: List[CursorOp]
+
+    override final def toString: String =
+      s"""History(${asList.mkString(", ")})"""
+  }
+
```

Comment:
Maybe add a comment about the typical direction a `History` stores the `CursorOps`, for instance, the `head` of the list is the first operation applied to the originating cursor and not the last (as it may have been previously)

Reply:
Yeah, I can do that. I struggled with what was the best way to both encode this and communicate this.